### PR TITLE
Bump `clap` to 3.0

### LINKED ISSUE #2920: Help text for `rustup update` contains a broken URL when line-broken
### Problem

Some terminals (including Windows Terminal) provide a mechanism for opening URLs in the default browser when you Ctrl-click on them. The help text for `rustup update` contains the URL `https://github.com/rust-lang/rustup/wiki/Non-host-toolchains`; however, if this gets broken across two lines the resulting URL is `https://github.com/rust` or `https://github.com/rust-`, which don't go where they're supposed to.

### Steps

1. Execute the command `rustup help update`.

### Possible Solution(s)

_No response_

### Notes

Obviously this requires a sufficiently small terminal window to word-wrap the help text.

### Rustup version

```console
rustup 1.24.3 (ce5817a94 2021-05-31)
```

### Installed toolchains

```console
Default host: x86_64-pc-windows-gnu
rustup home:  C:\Users\Thomas Winwood\.rustup

installed toolchains
--------------------

stable-x86_64-pc-windows-gnu (default)
nightly-x86_64-pc-windows-gnu

active toolchain
----------------

stable-x86_64-pc-windows-gnu (default)
rustc 1.57.0 (f1edd0429 2021-11-29)
```

### PR DESCRIPTION
**Pull request by zohnannor**
Bump `clap` to v3 which fixes incorrect executable name printing in subcommand's help texts on Windows (it comes as `rustup.exe-subcommand` in v2). I don't think this issue was reported to `rustup` but https://github.com/clap-rs/clap/pull/3693
Add `clap_complete` as some of `clap`'s functionality was moved into this crate.
Fix related tests.

close https://github.com/rust-lang/rustup/issues/2920

### CODE REVIEW

**Comment by kinnison** [comment]
This is a big chunk of change - have you been running this locally too to verify nothing bad happens?

**Comment by zohnannor** [comment]
I hadn't, but I will now, although I am not _using `rustup` directly_ as often (yes, I know, `cargo`, `rustc` and others are `rustup`'s proxies) besides updating `nightly` and `stable` Rust installations. I am not 100% sure that everything works besides the functionality that is tested using automatic tests, is there something that is not tested, that I should pay my attention to?

**Comment by zohnannor** [comment]
clap v4 got [released](https://github.com/clap-rs/clap/releases/tag/v4.0.4), might as well update to that version.

**Comment by 0xPoe** [comment]
> This is a big chunk of change - have you been running this locally too to verify nothing bad happens?

@kinnison What kind of tests do we need to do? I am glad to help review and test this PR.

File: src/cli/rustup_mode.rs
Comment (0xPoe) [inline src/cli/rustup_mode.rs]: Could you please explain why we need an append action here?

File: src/cli/rustup_mode.rs
Comment (0xPoe) [inline src/cli/rustup_mode.rs]: Why do we need nesting? I guess we only need one PossibleValuesParser.

File: src/cli/rustup_mode.rs
Comment (0xPoe) [inline src/cli/rustup_mode.rs]: ditto

**Comment by 0xPoe** [comment]
Thanks for working on this!

File: src/cli/rustup_mode.rs
Comment (zohnannor) [reply src/cli/rustup_mode.rs]: You are right, we don't. I was confused about `Arg::multiple`. Will fix other occurrences too.

File: src/cli/rustup_mode.rs
Comment (zohnannor) [reply src/cli/rustup_mode.rs]: Huh, some copy-paste error.

**Comment by zohnannor** [comment]
Still working on transition to `v4`, will be ready in a couple of days.

UPD: Forgot I can convert PR to draft.

**Comment by 0xPoe** [comment]
> Still working on transition to `v4`, will be ready in a couple of days.

I'd like to bump it up to v3 first and do more tests and release it. Because if we bump it up to v4, then we maybe get more broken changes.

File: src/cli/rustup_mode.rs
Comment (zohnannor) [inline src/cli/rustup_mode.rs]: Unsure about this: in `clap v2` there was [`FLAGS`, `OPTIONS`, `ARGS` sections, controlled by `clap`](https://github.com/clap-rs/clap/blob/master/CHANGELOG.md#highlights-1:~:text=You%20can%20now%20override%20the%20default%20group%20with%20Arg%3A%3Ahelp_heading%20and%20App%3A%3Asubcommand_help_heading.%20To%20apply%20a%20heading%20to%20a%20series%20of%20arguments%2C%20you%20can%20set%20App%3A%3Ahelp_heading.). Now, automatically added `-h, --help` and `-V, --version` flags are always displayed under `OPTIONS` section and there is no way to control that.

File: src/cli/rustup_mode.rs
Comment (zohnannor) [inline src/cli/rustup_mode.rs:1035]: I think this reads better than `m.try_get_one::<bool>("allow-downgrade").ok().flatten().copied().unwrap_or_default()`

And the reason we need `.try_` method here in the first place is because `update` function is used both for `rustup update` and `rustup toolchain install` that have different set of arguments.

File: src/cli/rustup_mode.rs
Comment (zohnannor) [inline src/cli/rustup_mode.rs]: In some places I think it's arguably better to still use `Arg::contains_id` for arguments that take value(s), instead of `.get_one::<String>().is_some()`.

UPD: changed my mind

File: src/cli/rustup_mode.rs
Comment (zohnannor) [inline src/cli/rustup_mode.rs]: Oops, this should be `.get_flag(name)` (as `DOCS_DATA` arguments are flags thanks to `.action(ArgAction::SetTrue)`), but it doesn't really matter too much.

File: src/cli/rustup_mode.rs
Comment (0xPoe) [reply src/cli/rustup_mode.rs]: Could you please test it and show the difference in the comment? Sorry for the delayed reply.

Comment (zohnannor) [reply src/cli/rustup_mode.rs]:
```diff 
-$ rustup --help
+$ ./target/debug/rustup --help
-rustup 1.25.1 (bb60b1e89 2022-07-12)
+rustup unknown (66def540e 2022-10-15) dirty 3 modifications
 The Rust toolchain installer
 
 USAGE:
-    rustup [FLAGS] [+toolchain] <SUBCOMMAND>
+    rustup [OPTIONS] [+toolchain] [SUBCOMMAND]
+
+OPTIONS:
+    -h, --help       Print help information
+    -V, --version    Print version information
 
 FLAGS:
     -v, --verbose    Enable verbose output
     -q, --quiet      Disable progress output
-    -h, --help       Prints help information
-    -V, --version    Prints version information
 
 ARGS:
     <+toolchain>    release channel (e.g. +stable) or custom toolchain to set override
@@ -31,7 +33,7 @@ SUBCOMMANDS:
     self           Modify the rustup installation
     set            Alter rustup settings
     completions    Generate tab-completion scripts for your shell
-    help           Prints this message or the help of the given subcommand(s)
+    help           Print this message or the help of the given subcommand(s)
 
 DISCUSSION:
     Rustup installs The Rust Programming Language from the official
@@ -43,33 +45,35 @@ DISCUSSION:
     If you are new to Rust consider running `rustup doc --book` to
     learn Rust.
 
-- snip --
```

<details>
<summary>Old vs New</summary>

<table>
<tr>
	<td> Old (`rustup 1.25 stable`) </td>
	<td> New (this `PR` 66def540) </td>
</tr>
<tr>
	<td> 

```
rustup 1.25.1 (bb60b1e89 2022-07-12)
The Rust toolchain installer

USAGE:
    rustup [FLAGS] [+toolchain] <SUBCOMMAND>

FLAGS:
    -v, --verbose    Enable verbose output
    -q, --quiet      Disable progress output
    -h, --help       Prints help information
    -V, --version    Prints version information

ARGS:
    <+toolchain>    release channel (e.g. +stable) or custom toolchain to set override

SUBCOMMANDS:
    show           Show the active and installed toolchains or profiles
    update         Update Rust toolchains and rustup
    check          Check for updates to Rust toolchains and rustup
    default        Set the default toolchain
    toolchain      Modify or query the installed toolchains
    target         Modify a toolchain's supported targets
    component      Modify a toolchain's installed components
    override       Modify directory toolchain overrides
    run            Run a command with an environment configured for a given toolchain
    which          Display which binary will be run for a given command
    doc            Open the documentation for the current toolchain
    man            View the man page for a given command
    self           Modify the rustup installation
    set            Alter rustup settings
    completions    Generate tab-completion scripts for your shell
    help           Prints this message or the help of the given subcommand(s)

DISCUSSION:
    Rustup installs The Rust Programming Language from the official
    release channels, enabling you to easily switch between stable,
    beta, and nightly compilers and keep them updated. It makes
    cross-compiling simpler with binary builds of the standard library
    for common platforms.
```

</td>
<td>


```
rustup unknown (66def540e 2022-10-15) dirty 3 modifications
The Rust toolchain installer

USAGE:
    rustup [OPTIONS] [+toolchain] [SUBCOMMAND]

OPTIONS:
    -h, --help       Print help information
    -V, --version    Print version information

FLAGS:
    -v, --verbose    Enable verbose output
    -q, --quiet      Disable progress output

ARGS:
    <+toolchain>    release channel (e.g. +stable) or custom toolchain to set override

SUBCOMMANDS:
    show           Show the active and installed toolchains or profiles
    update         Update Rust toolchains and rustup
    check          Check for updates to Rust toolchains and rustup
    default        Set the default toolchain
    toolchain      Modify or query the installed toolchains
    target         Modify a toolchain's supported targets
    component      Modify a toolchain's installed components
    override       Modify directory toolchain overrides
    run            Run a command with an environment configured for a given toolchain
    which          Display which binary will be run for a given command
    doc            Open the documentation for the current toolchain
    man            View the man page for a given command
    self           Modify the rustup installation
    set            Alter rustup settings
    completions    Generate tab-completion scripts for your shell
    help           Print this message or the help of the given subcommand(s)

DISCUSSION:
    Rustup installs The Rust Programming Language from the official
    release channels, enabling you to easily switch between stable,
    beta, and nightly compilers and keep them updated. It makes
    cross-compiling simpler with binary builds of the standard library
    for common platforms.

```

</td>
</tr>
</table>

</details>

**Comment by rbtcollins** [comment]
This doesn't bring joy :/

```

OPTIONS:
    -h, --help       Print help information
    -V, --version    Print version information

FLAGS:
    -v, --verbose    Enable verbose output
    -q, --quiet      Disable progress output
```

File: src/cli/rustup_mode.rs
Comment (0xPoe) [reply src/cli/rustup_mode.rs]: Cargo puts them all into the options. So I think we can follow it.

```console
➜  ~ cargo --help
Rust's package manager

Usage: cargo [+toolchain] [OPTIONS] [COMMAND]

Options:
  -V, --version             Print version info and exit
      --list                List installed commands
      --explain <CODE>      Run `rustc --explain CODE`
  -v, --verbose...          Use verbose output (-vv very verbose/build.rs output)
  -q, --quiet               Do not print cargo log messages
      --color <WHEN>        Coloring: auto, always, never
      --frozen              Require Cargo.lock and cache are up to date
      --locked              Require Cargo.lock is up to date
      --offline             Run without accessing the network
      --config <KEY=VALUE>  Override a configuration value
  -Z <FLAG>                 Unstable (nightly-only) flags to Cargo, see 'cargo -Z help' for details
  -h, --help                Print help information

Some common cargo commands are (see all commands with --list):
    build, b    Compile the current package
    check, c    Analyze the current package and report errors, but don't build object files
    clean       Remove the target directory
    doc, d      Build this package's and its dependencies' documentation
    new         Create a new cargo package
    init        Create a new cargo package in an existing directory
    add         Add dependencies to a manifest file
    remove      Remove dependencies from a manifest file
    run, r      Run a binary or example of the local package
    test, t     Run the tests
    bench       Run the benchmarks
    update      Update dependencies listed in Cargo.lock
    search      Search registry for crates
    publish     Package and upload this package to the registry
    install     Install a Rust binary. Default location is $HOME/.cargo/bin
    uninstall   Uninstall a Rust binary

See 'cargo help <command>' for more information on a specific command.
```

**Comment by jyn514** [comment]
If it helps, the clap migration guide recommends https://docs.rs/trycmd/latest/trycmd/ for testing the help output. To make it easier to see the differences, you could create a commit only adding trycmd, then upgrade clap, then bless the tests - the changes in the help output will show up in the diff, instead of having to reconstruct them from the code changes.

**Comment by rbtcollins** [comment]
I think that sort of test would be nice.

That said, the main thing is that the new UI will still be usable.

I think my main issue is that the cmd line help syntax line no longer matches the cmd line sections listed below it.

**Comment by 0xPoe** [comment]
> If it helps, the clap migration guide recommends https://docs.rs/trycmd/latest/trycmd/ for testing the help output. To make it easier to see the differences, you could create a commit only adding trycmd, then upgrade clap, then bless the tests - the changes in the help output will show up in the diff, instead of having to reconstruct them from the code changes.

@zohnannor Do you have time to add this test? If you are busy right now, I can help to add this test first. I think this test is not only useful for this time, but also we can use it in the future upgrade test.

**Comment by zohnannor** [comment]
@hi-rustin yes, pretty busy right now, I'll try to come back to this Soon™.

@jyn514 to clarify, you're suggesting to use `trycmd` just to make a diff between old and a new output and then remove `trycmd` or to integrate it in the project's tests suite and leave it there? I could do either. Before that, to see the output diff, I made a little script that wrote output of every `$ rustup <SUBCOMMAND> --help` to a file and used that on stable `rustup` and on the new one, then git-diff'ed it :^)

**Comment by jyn514** [comment]
> @jyn514 to clarify, you're suggesting to use `trycmd` just to make a diff between old and a new output and then remove `trycmd` or to integrate it in the project's tests suite and leave it there? I could do either. Before that, to see the output diff, I made a little script that wrote output of every `$ rustup <SUBCOMMAND> --help` to a file and used that on stable `rustup` and on the new one, then git-diff'ed it :^)

That's up to the rustup maintainers, but I'd personally recommend integrating it with the test suite and leaving it there, that will make https://github.com/rust-lang/rustup/pull/3094 easier too.

**Comment by 0xPoe** [comment]
> If you are busy right now, I can help to add this test first. I think this test is not only useful for this time, but also we can use it in the future upgrade test.



> @hi-rustin yes, pretty busy right now, I'll try to come back to this Soon™.

I am working on adding the trycmd tests.

File: src/cli/rustup_mode.rs
Comment (zohnannor) [reply src/cli/rustup_mode.rs]: So, I noticed they're using `Command::help_template` there, let's see how it would go after #3205 is merged.

**Comment by 0xPoe** [comment]
@zohnannor Friendly ping~  
Do you have any time to work on this PR? Because I was planning to do a release recently. So if we could get this change out that would be nice. Also, if you don't have time to work on this PR, I can help push your PR forward. I can squash your commits. And try to resolve the conflicts and fix all the tests, if you don't mind. I'm very sorry it took so long to go back and forth to get you to change this PR.

Thanks!  💚 💙 💜 💛 ❤️

**Comment by zohnannor** [comment]
Hey @hi-rustin thank you for caring!

>I'm very sorry it took so long to go back and forth to get you to change this PR.

No worries!

>I was planning to do a release recently.

May I ask for the deadline date?

I definitely will have time in the next couple (and by couple I mean ~4) of days to work on this again! I want to clarify some details of the changes to CLI API, the main complication for us is timezones :) So soon I'll compile a list of details and ask you to correct some of them if something is not as desired.

**Comment by rbtcollins** [comment]
I don't know when @hi-rustin was planning a release, but I can say that I have a thing that won't be ready before the weekend, that I really would love to include in the next release. So IMO the time frames are very compatible.

**Comment by 0xPoe** [comment]
> May I ask for the deadline date?

My original plan was to prepare a release in the first week of next month.

**Comment by 0xPoe** [comment]
> May I ask for the deadline date?

My original plan was to prepare a release in the first week of next month.

> I definitely will have time in the next couple (and by couple I mean ~4) of days to work on this again!

It's would be nice. Thank you very much. 

> So IMO the time frames are very compatible

Of course. Sounds good. I will wait for those changes. Thanks!

**Comment by zohnannor** [comment]
Hi. I got heavily sick, my plan of finishing this PR in the next couple of days is busted. @hi-rustin you can take and finish it like you said. Thanks for help, very sad I couldn't finish it myself. I still have a #3094 though :)

Sorry again for inconvenience.

**Comment by 0xPoe** [comment]
> Hi. I got heavily sick, my plan of finishing this PR in the next couple of days is busted.

Ah, I hope you get well soon.



> Sorry again for inconvenience.

No worry. Thanks for your contribution!

**Review by 0xPoe — APPROVED** [review]
Looks good to me!

After [this PR](https://github.com/rust-lang/rustup/pull/3252) lands, I'll retest it.

File: src/cli/rustup_mode.rs
Comment (rbtcollins) [inline src/cli/rustup_mode.rs:130]: This introduces a lot of noise.

What about

```
Ok(match matches.subcommand() {
  OK(s) => match s {
    ("dump-testament", _) => common::dump_testament()?,
     ...
     },
   _ => unreachable()
   }
```

**Review by rbtcollins — CHANGES_REQUESTED** [review]
this might need a rebase or something, some of the diff is a bit odd:


-        t.skip("tests/cli-ui/rustup/rustup_man_cmd_help_flag_stdout.toml");
+       t.skip("tests/suite/cli-ui/rustup/rustup_man_cmd_help_flag_stdout.toml");

**Comment by 0xPoe** [comment]
> this might need a rebase or something, some of the diff is a bit odd:

We just miss it last time. So I changed it in this PR. You can see https://github.com/rust-lang/rustup/blob/master/tests/suite/cli_ui.rs

File: src/cli/rustup_mode.rs
Comment (hi-rustin) [reply src/cli/rustup_mode.rs:130]: How about now?

Comment (0xPoe) [reply src/cli/rustup_mode.rs:130]: How about now?

**Review by rbtcollins — APPROVED** [review]
I'm curious about the new blank line after commands, might make for a little unneeded vertical whitespace on terminals, but nothing to prevent merging.

**Comment by 0xPoe** [comment]
I opened a question to ask the clap community. See https://github.com/clap-rs/clap/discussions/4745

**Review by 0xPoe — APPROVED** [review]
Check again

**Comment by 0xPoe** [comment]
I am going to merge this PR right now. @rbtcollins Thanks for your review!  @zohnannor Thanks for your contribution!

**Pull request merged**
